src/borg/testsuite/archiver.py

```diff
@@ -3766,35 +3764,23 @@ def get_changes(filename, data):
                 assert {'type': 'removed link'} in get_changes('input/link_removed', joutput)
 
                 # Symlink replacing or being replaced
-                if sys.platform == "darwin":
-                    # wierd: on macOS, symlink mode is 'lrwx------'. Is this true, or is there an
-                    # underlying problem here.
-                    assert {'type': 'mode', 'oldmode': "drwx------", 'newmode': 'lrwx------'} in \
-                        get_changes('input/dir_replaced_with_link', joutput)
-                    assert {'type': 'mode', 'oldmode': 'lrwx------', 'newmode': '-rw-------'} in \
-                        get_changes('input/link_replaced_by_file', joutput)
-                else:
-                    assert {'type': 'mode', 'oldmode': "drwx------", 'newmode': 'lrwxrwxrwx'} in \
-                        get_changes('input/dir_replaced_with_link', joutput)
-                    assert {'type': 'mode', 'oldmode': 'lrwxrwxrwx', 'newmode': '-rw-------'} in \
-                        get_changes('input/link_replaced_by_file', joutput)
+                # wierd: on macOS, symlink mode is 'lrwx------'. Is this true, or is there an underlying problem here.
+                expected_symlink_mode = 'lrwx------' if sys.platform == "darwin" else 'lrwxrwxrwx'
```

Comment:
I'll dig into this today, so we can answer that question.

Reply:
```
# i first created the target file, target_d dir, then created symlinks to them.
#
# first try, with umask 0022 (default on my system)
-rw-r--r--   1 vagrant  staff        0  3 Mär 07:30 target
drwxr-xr-x   2 vagrant  staff       68  3 Mär 07:29 target_d
lrwxr-xr-x   1 vagrant  staff        6  3 Mär 07:29 symlink -> target
lrwxr-xr-x   1 vagrant  staff        8  3 Mär 07:30 symlink_d -> target_d

# next try, with umask 0077 (manually set before invoking commands)
lrwx------   1 vagrant  staff        7  3 Mär 07:31 symlink2 -> target2
lrwx------   1 vagrant  staff        9  3 Mär 07:31 symlink_d2 -> target_d2
-rw-------   1 vagrant  staff        0  3 Mär 07:30 target2
drwx------   2 vagrant  staff       68  3 Mär 07:31 target_d2
```

So I suggest changing this test to not make it fail depending on what's set as umask.

E.g. result_mode.startswith('lrwx') could be used on both platforms (or even just 'l').

Reply:
also, remove the comment, it is clear now that this is expected.

if just testing for the first char, then the mode is not relevant at all, so no comment needed.

Reply:
I did a little digging, and it seems that linux always reports lrwxrwxrwx for symlinks but changes permissions for the pointed to file/directory when doing a chmod on a symlink (Python os.chmod(...,follow_symlinks=False) not allowed). macOS/bsd allows symlinks to have their own permissions. Not sure how that affets actual file access in case of conflicting permissions

Reply:
or should we make sure the permissions are set correctly when creating the test archives? If umask is causing different permissions to be set, then perhaps we should do an os.chmod to make sure permissions are what we want, so we actually test if borg detects the permission changes?

Reply:
here (https://superuser.com/questions/303040/how-do-file-permissions-apply-to-symlinks) is a good discussion on the macOS/bsd/linux symlink permissions. Haven't had time to digest yet - maybe this weekend.

Reply:
Slightly off-topic, but I just thought: if file is replaced by symlink, should there also be a size added/removed in the 'mode' changeset, indicating the removed size of the file and the added size of the symlink (length of the reference name)? Similarly for directory/link replacement and file/directory replacement, and maybe link/link replacement? Or should those have mutiple changesets, ie [{file removed}, {link added}, {mode changed}]? Currently just implemented logic as existing test mode used, but perhaps it should be changed (at least for json-lines mode, if we don't need them to be consistent).

Reply:
whether borg deals correctly with symlinks on all platforms is out of scope of this PR. you could open a separate issue with a todo to check this.

json diff just shows what we have in the archives / what changed, so that is unrelated.

also, for this PR it is expected to basically exactly do what the non-json mode is doing right now, everything else should be done later. open issues for later as needed.

aka "avoid scope creep".

Reply:
So, todo for here: mode.startswith(first_char)

Reply:
What I meant, wasn't whether borg properly backed-up and restored permissions, but was the fact that the test is creating files/directories/links without controlling permissions, and then is testing to see if borg did the right thing, but we don't know what the permissions actually are, and if they may change due to someone changing umask for example. Just thinking we should control the permissions and user/group when creating the files so that we properly test the features. I hope there is a test that verifies borg is correctly backing up permissions on the various platforms! As it stands, at least we know the files/directories are changing to/from symlinks. I may add another PR to improve the json-lines output info, but need to work on this PR for 1.2 and vorta 1st.

Reply:
if you like, you can also try to control and test the mode, but then it will be platform-specific as you already noticed.
but for now, i would also be happy with the less strict test, just testing for the file type change here.

Reply:
I pushed changes based on your recommendation, but the build failed, apparently for some system problem, not my changes. Can you restart the workflow, or do I need to force push something? Also, should I squash the PR into a few (or 1) more meaningful commits?

Reply:
I can squash all changes into 1 at the end.

Except if you prefer to do partial squashes and keep multiple commits - in that case please do that on your own.

Reply:
I just didn't want to squash the commits and have that throw off your reviewing. If you squash at merge that's fine - I just didn't want a lot of small reviewing-commits showing in the final merge.

Reply:
yeah, good! while reviewing squashing/force-pushing is rather annoying, little incremental changes are much better to review. i'll squash it at the end.